R/Solver.R

```diff
@@ -9,161 +9,197 @@
-        ### the first nullity(A') columns of which are an orthonormal basis for A',
-        ### so the nullity(F')-nullity(A') columns between are an orthonormal basis for F'\A'
-        kerFT_only <- qr.Q(qr(cbind(kerAT_basis,kerFT_basis),LAPACK=FALSE))[,(dim(kerAT_basis)[2]+1):(dim(kerFT_basis)[2]), drop=FALSE]
-        ### Re-normalize them to sum (in abs. value) to 1 and smooth near-zeros:
-        kerFT_norm <- apply(kerFT_only, MARGIN=2,
-                            FUN=function(col) ifelse(abs(col) < .Machine$double.eps, 0, col/sum(abs(col))))
-        
-        ## Get observation weights of these F'\A' basis vectors
-        ATw.weights <- t(A_mat) %*% kerFT_only
-        ### Re-normalize them to sum (in abs. value) to 1 and smooth near-zeros:
-        # ATw_norm <- apply(ATw.weights, MARGIN=2,
-        #                     FUN=function(col) ifelse(abs(col) < .Machine$double.eps, 0, col/sum(abs(col))))
-        ## Use normalized DID weights to get observation weights:
-        ATw_norm <- apply(t(A_mat) %*% kerFT_norm, MARGIN=2,
-                          FUN=function(col) ifelse(abs(col) < .Machine$double.eps, 0, col))
-        
-        ## Append F'\A' basis vector weights to Obs.weights and DID.weights as "Add.Obs.weights"
-        Obs.weights <- cbind(Obs.weights,Add.Obs.weights=ATw_norm)
-        DID.weights <- cbind(DID.weights,Add.Obs.weights=kerFT_norm)
-      } else if (FT_rank > RankAT) {
-        stop(simpleError("F has greater rank than A. Please check inputs for accuracy."))
-      }
-      
-      if (DID_full) { ## Add DID-estimator-only weights (that do not affect observation weights) if requested
-        print("Note: the weights labelled Add.DID.weights affect the estimator weights but not the observation weights.")
-        DID.weights <- cbind(DID.weights,Add.DID.weights=kerAT_norm)
-      } else { ## Otherwise print a note
-        print("Note: returning only a single solution for the DID estimator weights, althoughs others may exist that yield equivalent observation weights.")
-      }
-    }
-    return(list(DID.weights=DID.weights,
-                Obs.weights=Obs.weights))
+    stop("v must be a vector with length corresponding to the number of columns in F_mat, or a matrix of such column vectors.")
+  }
+  ## find a single solution for w:
+  w <- qr.solve(FT_qr, v)
+  if (is.null(colnames(w))) {
+    colnames(w) <- 1:ncol(w)
+  }
+
+  DID.weights <- data.frame(w)
+  colnames(DID.weights) <- paste("w.base", colnames(w), sep = ".")
+
+  ## find a single solution for weights for observations via A' * w:
+  Obs.weights <- data.frame(t(A_mat) %*% w)
+  colnames(Obs.weights) <- paste("ATw.base", colnames(w), sep = ".")
+
+  if (FT_rank == nrow(w)) { ## Prints note for single solution and jumps to return
+    print("There is a unique solution for the DID estimator weights.")
+    return(
+      list(
+        DID.weights = DID.weights,
+        Obs.weights = Obs.weights
+      ))
```

Comment:
Moved this `return()` statement inside the `if` condition to make it clear that we are returning a list after the `print()` statement.